File: svmgov/src/instructions/cast_vote_override.rs
use std::str::FromStr;

use anchor_client::solana_sdk::{
    instruction::Instruction, pubkey::Pubkey, signer::Signer, transaction::Transaction,
};
use anchor_lang::system_program;
use anyhow::{Result, anyhow};
use gov_v1::{ID as SNAPSHOT_PROGRAM_ID, MetaMerkleLeaf, MetaMerkleProof};
use log::info;

use crate::{
    constants::*,
    govcontract::client::{accounts, args},
    utils::{
        api_helpers::{
            convert_merkle_proof_strings, convert_stake_merkle_leaf_data_to_idl_type,
            generate_pdas_from_vote_proof_response, get_stake_account_proof,
            get_vote_account_proof,
        },
        utils::{
            create_spinner, derive_vote_override_cache_pda, derive_vote_override_pda,
            derive_vote_pda, setup_all_with_staker,
        },
    },
};

pub async fn cast_vote_override(
    proposal_id: String,
    ballot_id: u64,
    for_votes: u64,
    against_votes: u64,
    abstain_votes: u64,
    staker_keypair: String,
    rpc_url: Option<String>,
    _operator_api: Option<String>,
    stake_account_override: String,
    vote_account: String,
    snapshot_slot: u64,
    network: String,
) -> Result<()> {
    if for_votes + against_votes + abstain_votes != BASIS_POINTS_TOTAL {
        return Err(anyhow!(
            "Total vote basis points must sum to {}",
            BASIS_POINTS_TOTAL
        ));
    }

    let proposal_pubkey = Pubkey::from_str(&proposal_id)
        .map_err(|_| anyhow!("Invalid proposal ID: {}", proposal_id))?;

    let (payer, program, merkle_proof_program) = setup_all_with_staker(staker_keypair, rpc_url)?;

    let stake_account_str = stake_account_override.clone();
    let vote_account_pubkey = Pubkey::from_str(&vote_account)
        .map_err(|_| anyhow!("Invalid vote account: {}", vote_account))?;

    let meta_merkle_proof = get_vote_account_proof(&vote_account, snapshot_slot, &network).await?;

    let stake_merkle_proof =
        get_stake_account_proof(&stake_account_str, snapshot_slot, &network).await?;

    let (consensus_result_pda, meta_merkle_proof_pda) =
        generate_pdas_from_vote_proof_response(ballot_id, &meta_merkle_proof)?;

    let validator_vote_pda = derive_vote_pda(&proposal_pubkey, &vote_account_pubkey, &program.id());
    let vote_override_pda = derive_vote_override_pda(
        &proposal_pubkey,
        &Pubkey::from_str(&stake_account_str)?,
        &validator_vote_pda,
        &program.id(),
    );
    let vote_override_cache_pda =
        derive_vote_override_cache_pda(&proposal_pubkey, &validator_vote_pda, &program.id());

    let stake_merkle_proof_vec =
        convert_merkle_proof_strings(&stake_merkle_proof.stake_merkle_proof)?;

    let stake_merkle_leaf =
        convert_stake_merkle_leaf_data_to_idl_type(&stake_merkle_proof.stake_merkle_leaf)?;

    let spinner = create_spinner("Sending vote override transaction...");

    let mut ixs: Vec<Instruction> = vec![];

    // Check if meta merkle proof account exists, create if missing
    let meta_merkle_proof_account = match program
        .account::<MetaMerkleProof>(meta_merkle_proof_pda)
        .await
    {
        Ok(account) => Some(account),
        Err(_e) => {
            info!("Unable to get meta merkle proof account, will create it");
            None
        }
    };

    if meta_merkle_proof_account.is_none() {
        info!("Creating meta merkle proof account");

        let voting_wallet = Pubkey::from_str(&meta_merkle_proof.meta_merkle_leaf.voting_wallet)
            .map_err(|e| anyhow!("Invalid voting wallet in proof: {}", e))?;

        let init_meta_merkle_proof_ix = merkle_proof_program
            .request()
            .args(gov_v1::instruction::InitMetaMerkleProof {
                close_timestamp: 1,
                meta_merkle_leaf: MetaMerkleLeaf {
                    voting_wallet,
                    vote_account: vote_account_pubkey,
                    stake_merkle_root: Pubkey::from_str_const(
                        meta_merkle_proof
                            .meta_merkle_leaf
                            .stake_merkle_root
                            .as_str(),
                    )
                    .to_bytes(),
                    active_stake: meta_merkle_proof.meta_merkle_leaf.active_stake,
                },
                meta_merkle_proof: meta_merkle_proof
                    .meta_merkle_proof
                    .iter()
                    .map(|s| Pubkey::from_str_const(s).to_bytes())
                    .collect(),
            })
            .accounts(gov_v1::accounts::InitMetaMerkleProof {
                consensus_result: consensus_result_pda,
                merkle_proof: meta_merkle_proof_pda,
                payer: payer.pubkey(),
                system_program: system_program::ID,
            })
            .instructions()?;

        ixs.extend(init_meta_merkle_proof_ix);
    }

    let cast_vote_override_ixs = program
        .request()
        .args(args::CastVoteOverride {
            for_votes_bp: for_votes,
            against_votes_bp: against_votes,
            abstain_votes_bp: abstain_votes,
            stake_merkle_proof: stake_merkle_proof_vec,
            stake_merkle_leaf,
        })
        .accounts(accounts::CastVoteOverride {
            signer: payer.pubkey(),
            spl_vote_account: vote_account_pubkey,
            spl_stake_account: Pubkey::from_str(&stake_account_str)?,
            proposal: proposal_pubkey,
            validator_vote: validator_vote_pda,
            vote_override: vote_override_pda,
            vote_override_cache: vote_override_cache_pda,
            consensus_result: consensus_result_pda,
            meta_merkle_proof: meta_merkle_proof_pda,
            snapshot_program: SNAPSHOT_PROGRAM_ID,
            system_program: system_program::ID,
        })
        .instructions()?;

    ixs.extend(cast_vote_override_ixs);

    let blockhash = program.rpc().get_latest_blockhash().await?;
    let transaction =
        Transaction::new_signed_with_payer(&ixs, Some(&payer.pubkey()), &[&payer], blockhash);

    let sig = program
        .rpc()
        .send_and_confirm_transaction(&transaction)
        .await?;

    spinner.finish_with_message(format!(
        "Vote override cast successfully. https://explorer.solana.com/tx/{}",
        sig
    ));

    Ok(())
}

File: frontend/src/chain/instructions/castVoteOverride.ts
import { PublicKey, SystemProgram, Transaction } from "@solana/web3.js";
import {
  CastVoteOverrideParams,
  TransactionResult,
  BlockchainParams,
  SNAPSHOT_PROGRAM_ID,
} from "./types";
import {
  createProgramWithWallet,
  createGovV1ProgramWithWallet,
  getVoteAccountProof,
  getStakeAccountProof,
  convertMerkleProofStrings,
  convertStakeMerkleLeafDataToIdlType,
  validateVoteBasisPoints,
  deriveVotePda,
  deriveVoteOverridePda,
  deriveVoteOverrideCachePda,
  getMetaMerkleProofPda,
} from "./helpers";
import { BN } from "@coral-xyz/anchor";

/**
 * Casts a vote override using a stake account
 */
export async function castVoteOverride(
  params: CastVoteOverrideParams,
  blockchainParams: BlockchainParams,
  slot: number | undefined
): Promise<TransactionResult> {
  const {
    proposalId,
    forVotesBp,
    againstVotesBp,
    abstainVotesBp,
    stakeAccount,
    wallet,
    voteAccount,
    consensusResult,
  } = params;

  if (!wallet || !wallet.publicKey) {
    throw new Error("Wallet not connected");
  }

  if (slot === undefined) {
    throw new Error("Slot is not defined");
  }

  if (consensusResult === undefined) {
    throw new Error("Consensus result not defined");
  }

  // Validate vote distribution
  validateVoteBasisPoints(forVotesBp, againstVotesBp, abstainVotesBp);

  const proposalPubkey = new PublicKey(proposalId);
  const splVoteAccount = new PublicKey(voteAccount);
  const program = createProgramWithWallet(wallet, blockchainParams.endpoint);
  const snapshotProgram = createGovV1ProgramWithWallet(
    wallet,
    blockchainParams.endpoint
  );

  const stakeAccountPubkey = new PublicKey(stakeAccount);

  const consensusAcc = await snapshotProgram.account.consensusResult.fetch(
    consensusResult
  );
  console.log(
    "consensusAcc:",
    new PublicKey(consensusAcc.ballot.metaMerkleRoot).toBase58(),
    new PublicKey(consensusAcc.ballot.snapshotHash).toBase58()
  );

  // consensusAcc: BJC3zN8wq1GeCBFCLf5rQXo3QLFZbyfLJXWaJvrzMx6Q CfzhKPXFe7YahXVmzpKRyaJqxAsvz214Zo2VFZJMsioj

  // '{"snapshotSlot":"15f1c98a",
  // "ballot":{"metaMerkleRoot":[152,252,61,48,248,172,236,241,130,20,144,239,95,110,138,194,255,246,118,0,71,109,109,64,93,151,250,6,221,77,204,5],
  // "snapshotHash":[173,109,220,238,165,65,12,2,88,69,143,103,211,137,77,187,159,85,207,156,227,119,49,68,41,70,154,21,105,98,175,10]},
  // "tieBreakerConsensus":false}'

  // SendTransactionError: Simulation failed.
  // Message: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x1779.
  // Logs:
  // [
  //   "Program DYvhrdFv8PSzta2u2igEoAXR6A6nzUXBnVhy6iNM6F7o invoke [1]",
  //   "Program log: Instruction: InitMetaMerkleProof",
  //   "Program 11111111111111111111111111111111 invoke [2]",
  //   "Program 11111111111111111111111111111111 success",
  //   "Program log: Root BJC3zN8wq1GeCBFCLf5rQXo3QLFZbyfLJXWaJvrzMx6Q != Node 9yFC99swZereuzurg6Wy9VYM1tGC5v5AbvQJMk2WYcji",
  //   "Program log: AnchorError thrown in programs/gov-v1/src/merkle_helper.rs:48. Error Code: InvalidMerkleProof. Error Number: 6009. Error Message: Invalid merkle proof.",
  //   "Program DYvhrdFv8PSzta2u2igEoAXR6A6nzUXBnVhy6iNM6F7o consumed 20113 of 200000 compute units",
  //   "Program DYvhrdFv8PSzta2u2igEoAXR6A6nzUXBnVhy6iNM6F7o failed: custom program error: 0x1779"
  // ].

  // Get proofs
  const network = blockchainParams.network;
  const [metaMerkleProof, stakeMerkleProof] = await Promise.all([
    getVoteAccountProof(splVoteAccount.toBase58(), network, slot),
    getStakeAccountProof(stakeAccount, network, slot),
  ]);

  const metaMerkleProofPda = getMetaMerkleProofPda(
    metaMerkleProof,
    SNAPSHOT_PROGRAM_ID,
    consensusResult
  );

  // Check if merkle account exists
  const merkleAccountInfo = await program.provider.connection.getAccountInfo(
    metaMerkleProofPda,
    "confirmed"
  );

  if (!merkleAccountInfo) {
    const govV1Program = createGovV1ProgramWithWallet(
      wallet,
      blockchainParams.endpoint
    );

    const stakeMerkleRootData = Array.from(
      new PublicKey(
        metaMerkleProof.meta_merkle_leaf.stake_merkle_root
      ).toBytes()
    );

    const metaMerkleProofData = metaMerkleProof.meta_merkle_proof.map((proof) =>
      Array.from(new PublicKey(proof).toBytes())
    );

    console.log("log:", {
      consensusResult,
      merkleProof: metaMerkleProofPda,
      payer: wallet.publicKey,
      systemProgram: SystemProgram.programId,
    });

    // '{"consensusResult":"AagyvVs676jmytwMUPtwefhynBhhJ7zir57Mf4vuu2UL",
    // "merkleProof":"EV458DMNTaRjxmAhuPrw1QocjrcWAdmW36ZjkTq2GkU1",
    // "payer":"2qr3GM59WXRi2hPkQ9ATugpqMG6p4YFsuSLYYLSzQfGT","systemProgram":"11111111111111111111111111111111"}'

    const initMerkleInstruction = await govV1Program.methods
      .initMetaMerkleProof(
        {
          votingWallet: new PublicKey(
            metaMerkleProof.meta_merkle_leaf.voting_wallet
          ),
          voteAccount: new PublicKey(
            metaMerkleProof.meta_merkle_leaf.vote_account
          ),
          stakeMerkleRoot: stakeMerkleRootData,
          activeStake: new BN(
            `${metaMerkleProof.meta_merkle_leaf.active_stake}`
          ),
        },
        metaMerkleProofData,
        new BN(1)
      )
      .accountsStrict({
        consensusResult,
        merkleProof: metaMerkleProofPda,
        payer: wallet.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .instruction();

    const recentBlockhash =
      await program.provider.connection.getLatestBlockhash("confirmed");

    const transaction = new Transaction();
    transaction.add(initMerkleInstruction);
    transaction.feePayer = wallet.publicKey;
    transaction.recentBlockhash = recentBlockhash.blockhash;

    const tx = await wallet.signTransaction(transaction);

    const signature = await program.provider.connection.sendRawTransaction(
      tx.serialize(),
      { preflightCommitment: "confirmed" }
    );
    console.log("signature initMerkleInstruction", signature);
  }

  // Convert merkle proof data
  const stakeMerkleProofVec = convertMerkleProofStrings(
    stakeMerkleProof.stake_merkle_proof
  );
  const stakeMerkleLeaf = convertStakeMerkleLeafDataToIdlType(
    stakeMerkleProof.stake_merkle_leaf
  );

  const forVotesBn = new BN(forVotesBp);
  const againstVotesBn = new BN(againstVotesBp);
  const abstainVotesBn = new BN(abstainVotesBp);

  const votePda = deriveVotePda(
    proposalPubkey,
    splVoteAccount,
    program.programId
  );

  const voteOverridePda = deriveVoteOverridePda(
    proposalPubkey,
    stakeAccountPubkey,
    votePda,
    program.programId
  );

  const voteOverrideCachePda = deriveVoteOverrideCachePda(
    proposalPubkey,
    votePda,
    program.programId
  );

  // Build cast vote override instruction
  const castVoteOverrideInstruction = await program.methods
    .castVoteOverride(
      forVotesBn,
      againstVotesBn,
      abstainVotesBn,
      stakeMerkleProofVec,
      stakeMerkleLeaf
    )
    .accountsStrict({
      signer: wallet.publicKey,
      splVoteAccount: splVoteAccount,
      splStakeAccount: stakeAccountPubkey,
      proposal: proposalPubkey,
      consensusResult,
      metaMerkleProof: metaMerkleProofPda,
      snapshotProgram: SNAPSHOT_PROGRAM_ID,
      systemProgram: SystemProgram.programId,
      voteOverride: voteOverridePda,
      voteOverrideCache: voteOverrideCachePda,
      validatorVote: votePda,
    })
    .instruction();

  const transaction = new Transaction();
  transaction.add(castVoteOverrideInstruction);
  transaction.feePayer = wallet.publicKey;
  transaction.recentBlockhash = (
    await program.provider.connection.getLatestBlockhash("confirmed")
  ).blockhash;

  const tx = await wallet.signTransaction(transaction);

  const signature = await program.provider.connection.sendRawTransaction(
    tx.serialize(),
    { preflightCommitment: "confirmed" }
  );
  console.log("signature cast vote override", signature);
  return {
    signature,
    success: true,
  };
}
